url2698a2c731_cassandra_project_cassandra_stand_alone_tgz-pJ8-eu_cassandra_sim_MySimpleDynamicScenarioSimulationJ8/src/eu/cassandra/sim/Simulation.java

```diff
@@ -0,0 +1,1042 @@
+		return endTick;
+	}
+
+	/**
+	 * Gets the random number generator.
+	 *
+	 * @return the orng
+	 */
+	public ORNG getOrng() {
+		return orng;
+	}
+	
+	/**
+	 * Gets the simulation parameters set.
+	 *
+	 * @return the simulation world
+	 */
+	public SimulationParams getSimulationWorld () {
+		return simulationWorld;
+	}
+
+	
+	/**
+	 * Prints the aggregate KPIs, followed by the KPIs for all installations, appliances and activities.
+	 */
+	public void printKPIs()
+	{
+		System.out.println();
+		printKPIs(m.getKPIs(DBResults.AGGR), null, DBResults.AGGR);
+		for (Installation inst: this.getInstallations())
+		{
+			printKPIs(m.getKPIs(inst.getId()), "installation", inst.getId());
+			for (Appliance app: inst.getAppliances())
+				printKPIs(m.getAppKPIs(app.getId()), "appliance", app.getId());
+			for (Person p: inst.getPersons())
+				for (Activity a: p.getActivities())
+					printKPIs(m.getActKPIs(a.getId()), p.getName() + "'s activity", a.getId());
+		}
+	}
+
+	protected void printKPIs( HashMap<String, Double> kpis, String entityType, String entityId)
+	{
+		if (entityType == null && entityId == DBResults.AGGR)
+			System.out.println("Aggregate KPIs");
+		else
+			System.out.println("KPIs for " + entityType + " " + entityId);
+		for (String key: kpis.keySet())
+			System.out.println(key + " \t" + kpis.get(key));
+		System.out.println();
+	}
+
+
+	private void calculateExpectedPower(String dbname) {
+		double[] aggr_exp = new double[Constants.MIN_IN_DAY];
+		for(Installation installation: installations) {
+			double[] inst_exp = new double[Constants.MIN_IN_DAY];
+			Person person = installation.getPersons().get(0);
+			for(Activity activity: person.getActivities()) {
+				double[] act_exp = activity.calcExpPower();
+				for(int i = 0; i < act_exp.length; i++) {
+					inst_exp[i] += act_exp[i];
+					m.addExpectedPowerTick(i, activity.getId(), act_exp[i], DBResults.COL_ACTRESULTS_EXP);
+				}
+			}
+			// For every appliance that is a base load find mean value and add
+			for(Appliance appliance: installation.getAppliances()) {
+				if(appliance.isBase()) {
+					double mean = 0;
+					Double[] cons = appliance.getActiveConsumption();
+					for(int i = 0; i < cons.length; i++) {
+						mean += cons[i].doubleValue();
+					}
+					mean /= cons.length;
+					for(int i = 0; i < inst_exp.length; i++) {
+						inst_exp[i] += mean;
+					}
+				}
+			}
+
+			for(int i = 0; i < inst_exp.length; i++) {
+				aggr_exp[i] += inst_exp[i];
+				m.addExpectedPowerTick(i, installation.getId(), inst_exp[i], DBResults.COL_INSTRESULTS_EXP);
+			}
+		}
+		for(int i = 0; i < aggr_exp.length; i++) 
+			m.addExpectedPowerTick(i, "aggr", aggr_exp[i], DBResults.COL_AGGRRESULTS_EXP);
+
+		// Write installation results to csv file
+		if (!resources_path.endsWith("/"))
+			resources_path += "/";
+		String filename = resources_path + dbname + "_exp_pow.csv";
+		File csvFile = new File(filename);
+		FileWriter fw;
+		try {
+			fw = new FileWriter(csvFile);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *fw*. The resource is referenced at lines 947, 975, 977. The resource is closed at line 978. There are other execution paths that do not contain closure statements, for example, when OutputStreamWriter.flush throws an exception. Ensure *fw* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).